complete the implementation of __make_ancestors_values_stages and __m…

Gczmy opened the pull request:
## Main changes

Complete the implementation of `__make_ancestors_values_stages` and `__make_probability_values`.

Gczmy committed:
complete the implementation of __make_ancestors_values_stages and __make_probability_values

Gczmy committed:
resolve conflicts

alphaville commented on raocp/core/scenario_tree.py:
```diff
@@ -106,22 +114,39 @@ def __make_ancestors_values_stages(self):
+        probs[0] = 1
+        probs[1:] = self.__initial_distribution[np.flatnonzero(self.__initial_distribution)]
+        for i in range(num_nonzero_init_distr+1, len(values)):
+            probs_new = probs[int(ancestors[i])] * \
```
these `int` are not necessary

alphaville commented on raocp/core/scenario_tree.py:
```diff
@@ -106,22 +114,39 @@ def __make_ancestors_values_stages(self):
+        :return: probability
+        """
+        num_nonzero_init_distr = len(list(filter(lambda x: (x > 0), self.__initial_distribution)))
+        ancestors, values, stages = self.__make_ancestors_values_stages()
```
`self.__make_ancestors_values_stages` is a time consuming method that has been already invoked; you don't need to call it again

alphaville commented on raocp/core/scenario_tree.py:
```diff
@@ -106,22 +114,39 @@ def __make_ancestors_values_stages(self):
+
         return ancestors, values, stages
 
     def __make_probability_values(self):
```
This implementation doesn't take into account the fact that there is a stopping time after which there is no more branching. To see this, add this line at the end of `test_markov_probability_of_node` in `tests/test_scenario_tree.py`:

```.py
self.assertAlmostEqual(0.005, tree.probability_of_node(20), delta=tol)
```

alphaville commented on raocp/core/scenario_tree.py:
```diff
@@ -53,13 +59,15 @@ def value_at_node(self, node_idx):
+        return self.children_of(self.ancestor_of(node_idx))
 
     def conditional_probabilities_of_children(self, node_idx):
         raise NotImplementedError()
```
try this:

```.py
def conditional_probabilities_of_children(self, node_idx):
    prob_node_idx = self.probability_of_node(node_idx)
    children = self.children_of(node_idx)
    prob_children = self.__probability[children]
    return prob_children/prob_node_idx
```

alphaville commented on raocp/core/scenario_tree.py:
```diff
@@ -106,22 +114,39 @@ def __make_ancestors_values_stages(self):
+        probs[1:] = self.__initial_distribution[np.flatnonzero(self.__initial_distribution)]
+        for i in range(num_nonzero_init_distr+1, len(values)):
+            probs_new = probs[int(ancestors[i])] * \
+                        self.__transition_prob[int(values[int(ancestors[i])]), int(values[i])] * np.ones(1, )
```
I'm not sure what `np.ones(1, )` does

alphaville committed:
Merge branch 'dev/scenario-tree' into dev/__make_ancestors_values_stages

alphaville committed:
update make probability

- implementation of conditional probabilities of children
- fix probability values after stopping time
- all tests but one pass

alphaville committed:
all tests now pass
